### ISSUE TITLE
Expose effective dispatcher parallelism

### ISSUE THREAD

**Issue by dovchinnikov** [Kotlin/kotlinx.coroutines#3997]
**Use case**

I'm writing a bunch of utilities, which accept `concurrency` parameter:
- default concurrency should be the same as dispatcher effective parallelism.
- user-specified concurrency should be capped by dispatcher effective parallelism.

Example:
```
const val DEFAULT_CONCURRENCY = -1
const val SEQUENTIAL = 1

suspend fun Collection<T>.forEachWithProgress(concurrency: Int = SEQUENTIAL, action: suspend (T) -> Unit) {
  val effectiveConcurrency = when {
    concurrency  == DEFAULT_CONCURRENCY ->  `current dispatcher effective concurrency`
    concurrency > 0 -> min(concurrency, `current dispatcher effective concurrency`)
    else -> error("unsupported $concurrency")
  }
  if (effectiveConcurrency == SEQUENTIAL) {
    // no coroutines here, just a for loop
  } else {
    // create a channel + effectiveConcurrency coroutines to handle this collection
  }
}
```

**Comment by dkhalanskyjb** [comment]
If the dispatcher has, say, two threads (so you launch at most two coroutines) and `action` suspends for a long time, these threads will lie dormant, even though they could potentially do some work.

Is this for the sake of performance? If so, could you share any benchmarks that show that exceeding `current dispatcher effective concurrency` leads to measurable performance regressions? I honestly have no intuition how much better (or worse) this is than some reasonable problem-area-specific constant in (16..128).

**Comment by dovchinnikov** [comment]
> If the dispatcher has, say, two threads (so you launch at most two coroutines) and action suspends for a long time, these threads will lie dormant, even though they could potentially do some work.

No, why would it lie dormant? It will process other tasks in its queue. 

> Is this for the sake of performance? 

Yes. We've had cases when the collection has couple million elements, and before that we've created a coroutine-per-element in order to let the dispatcher arrange the execution at its own discretion. We've discovered that creating a million coroutines is cheap but not free (I'm sorry I cannot provide a benchmark or snapshot). It's much cheaper to run a number of coroutines equal to CPU size, and let them process the channel concurrently. So, instead of N coroutines, we now have 1 producer coroutine which sends elements from the collection to the Channel, and CPU-count coroutines which are receiving elements from the Channel. This became way faster! (I guess because doing nothing a million times is faster than doing something a million times.)

Now, why CPU count? It's effectively a maximum physical parallelism. But ideally I want to run as much coroutines as possible to fit in a given dispatcher, i.e. why even run coroutines are create a channel when the dispatcher is effectively sequential (`runBlocking`/`limitedParallelism(1)`)? The point to avoid unnecessary computations, e.g. allocating CPU=24 coroutines on a sequential dispatcher. Also, supporting a generic API requires some user input validation, and I want to disallow specifying too big `concurrency` for no reason.

**Comment by dovchinnikov** [comment]
Please note that the effective parallelism already exists. I was going to expose it in https://github.com/Kotlin/kotlinx.coroutines/commit/b74e039196711512764add6f94af92f99d09d3aa but there was no use case for it, and now there is one

**Comment by dkhalanskyjb** [comment]
> No, why would it lie dormant? It will process other tasks in its queue.

If there are any, sure. Even in that case, you could saturate its queue more fully with just `forEachWithProgress` if concurrency exceeds parallelism.

> We've discovered that creating a million coroutines is cheap but not free (I'm sorry I cannot provide a benchmark or snapshot). It's much cheaper to run a number of coroutines equal to CPU size

This agrees with my intuition entirely! What's not clear to me is that there isn't a universally good number like 64 between |CPU| and one million that's indistinguishable from |CPU| performance-wise.

> why even run coroutines are create a channel when the dispatcher is effectively sequential

For any use case where a coroutine can suspend for a long time, and therefore, concurrency exceeds parallelism.

```kotlin
launch(Dispatchers.Default.limitedParallelism(1)) {
  urls.forEachWithProgress(concurrency = 10) {
      val page = withContext(Dispatchers.IO) { getPage(it) }
      val data = compute(page)
      withContext(Dispatchers.IO) { storeDataInDb(data) }
  }
```

This way, there are at most 10 outstanding connections, but at most a single thread busy with computations.

**Comment by fvasco** [comment]
> if (effectiveConcurrency == SEQUENTIAL) {
>     // no coroutines here, just a for loop
>   } else {
>     // create a channel + effectiveConcurrency coroutines to handle this collection
>   }

I suppose there isn't any suspension point inside the `SEQUENTIAL` block.
Please consider that a single thread dispatcher _can_ execute multiple coroutines _in parallel_.

**Comment by dovchinnikov** [comment]
> there isn't a universally good number .. between |CPU| and one million that's indistinguishable from |CPU| performance-wise

Do you mean I should keep the current solution with |CPU| as `DEFAULT_CONCURRENCY`? 

> This way, there are at most 10 outstanding connections, but at most a single thread busy with computations.

You are correct. I can see that it's okay if concurrency is greater than effective dispatcher parallelism.

> I suppose there isn't any suspension point inside the SEQUENTIAL block.

There is a `yield` before each element. 

> Please consider that a single thread dispatcher can execute multiple coroutines in parallel.

A single thread dispatcher can execute multiple coroutines concurrently, but there will be at most 1 coroutine actually running at any given time.

**Comment by dkhalanskyjb** [comment]
> Do you mean I should keep the current solution with |CPU| as `DEFAULT_CONCURRENCY`?

Maybe. Or maybe something like `minOf(size, maxOf(|CPU| * 2, 100))`. Or `minOf(size, 500)`. There are several orders of magnitude between `|CPU|` and a million, and the optimal solution for your payload could easily be somewhere between them.

There is advice of this form (though the number varies) circulating the public consciousness: "If you don't know how your thread pool will be used, having |CPU| * 2 + 1 threads by default is a good rule of thumb. This way, when threads block, the CPU will still probably be efficiently utilized." This is reasonable: the number is low enough that we don't risk degrading performance by spawning too many threads, but high enough that, even when some threads block, the CPU is highly utilized.

If we adopt advice like this for coroutines, we get: "If you don't know how your <s>thread pool</s> set of coroutines will be used, having <s>|CPU|</s> `effective concurrency` * 2 + 1 <s>threads</s> coroutines by default is a good rule of thumb. This way, when <s>threads block</s> coroutines suspend, the <s>CPU</s> OS threads will still probably be efficiently utilized." But I don't think we should adopt this advice verbatim. A major point of coroutines is that they are lightweight. It'll probably noticeably affect performance when you overshoot the number of created threads by 100, so sticking to `c * |CPU|` makes sense, but when you overshoot the number of created coroutines by 100, is this really ever going to show up in the profiler other than for corner-case things like `{ x -> Unit }`? Erring on the side of spawning too many coroutines doesn't seem as problematic as spawning too few.

If it does turn out that keeping the number of coroutines close to `effective concurrency` brings statistically significant benefits, then sure, this seems like a valuable API to add. But I would certainly be surprised.

**Comment by dovchinnikov** [comment]
I've decided to go with `minOf(size, |CPU|)` by default since most of usages do CPU-bound work without switching to another dispatcher.
